### PR TITLE
Add support for Deployments and Environments api

### PR DESCRIPTION
**Pull request by zegenie**
Added basic support for fetching information about Deployments and Environments, since we needed this information in one of our deploy tools.

### CODE REVIEW

File: lib/Gitlab/Api/Environments.php [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+    {
+        return $this->get($this->getProjectPath($project_id, 'environments'));
+    }
+
+    /**
+     * @param int $project_id
+     * @param array $params
+     * @return mixed
+     */
+    public function create($project_id, array $params = array())
```
Comment (m1guelpf) [inline lib/Gitlab/Api/Environments.php]: Could you use OptionsResolver here?

Comment (zegenie) [reply lib/Gitlab/Api/Environments.php]: I was looking at other `create()` methods and I can't see any using the `OptionsResolver` - was just going with what seemed like the convention. Will have a look and update if possible.

**Comment by zegenie** [comment]
Updated with using `OptionsResolver` for resolving `create()` parameters

File: lib/Gitlab/Api/Deployments.php [hunk 2]
```diff
@@ -0,0 +1,23 @@
+<?php namespace Gitlab\Api;
+
+class Deployments extends AbstractApi
+{
+    /**
+     * @param int $project_id
+     * @return mixed
+     */
+    public function all($project_id)
```
Comment (fbourigault) [inline lib/Gitlab/Api/Deployments.php]: Could use the `OptionsResolver` like in https://github.com/m4tthumphrey/php-gitlab-api/blob/master/lib/Gitlab/Api/Repositories.php#L13 to allow pagination?

File: lib/Gitlab/Api/Environments.php [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+
+use Symfony\Component\OptionsResolver\OptionsResolver;
+
+class Environments extends AbstractApi
+{
+    /**
+     * @param int $project_id
+     * @return mixed
+     */
+    public function all($project_id)
```
Comment (fbourigault) [inline lib/Gitlab/Api/Environments.php]: Same here for pagination.

File: lib/Gitlab/Api/Environments.php [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+
+    /**
+     * @param int $project_id
+     * @param array $parameters
+     * @return mixed
+     */
+    public function create($project_id, array $parameters = array())
+    {
+        $resolver = new OptionsResolver();
+        $resolver->setDefined('name')
```
Comment (fbourigault) [inline lib/Gitlab/Api/Environments.php]: According to the documentation this parameter is required. Could you use `setRequired` instead?

File: lib/Gitlab/Api/Environments.php [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+            ->setAllowedTypes('name', 'string');
+        $resolver->setDefined('slug')
```
Comment (fbourigault) [inline lib/Gitlab/Api/Environments.php]: I don't find any reference in documentation about this `slug` argument. Could you provide some official input that prove it exists? I used to reject parameters not officially documented as those may be painful to maintain in the future.

File: lib/Gitlab/Api/Environments.php [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+     * @return mixed
+     */
+    public function all($project_id)
+    {
+        return $this->get($this->getProjectPath($project_id, 'environments'));
+    }
+
+    /**
+     * @param int $project_id
+     * @param array $parameters
```
Comment (fbourigault) [inline lib/Gitlab/Api/Environments.php]: Could you add details about this argument like in:
https://github.com/m4tthumphrey/php-gitlab-api/blob/35ecf9b82984beace5eaea0df4fcbe4f77e2c285/lib/Gitlab/Api/Repositories.php#L179-L194

File: lib/Gitlab/Api/Environments.php (hunk 4 above)
Comment (zegenie) [reply lib/Gitlab/Api/Environments.php]: Done.

File: lib/Gitlab/Api/Deployments.php (hunk 2 above)
Comment (zegenie) [reply lib/Gitlab/Api/Deployments.php]: Done. I wasn't sure whether this had pagination so I just left it out. I also added `$parameter` in the docblock for `Repositories` used as example as it was missing.

File: lib/Gitlab/Api/Environments.php (hunk 3 above)
Comment (zegenie) [reply lib/Gitlab/Api/Environments.php]: Done.

File: lib/Gitlab/Api/Environments.php (hunk 6 above)
Comment (zegenie) [reply lib/Gitlab/Api/Environments.php]: Sure thing. Done, using github API description for the details.

File: lib/Gitlab/Api/Environments.php (hunk 5 above)
Comment (zegenie) [reply lib/Gitlab/Api/Environments.php]: I removed this as I had just misunderstood the return value coming back from the api. It was never a valid parameter. Good catch.

**Comment by fbourigault** [comment]
Thank you!

**Pull request merged**